File: python/49.group-anagrams.py
class Solution(object):
    def solve1(self, strs):
        groups = {}
        for _str in strs:
            _sorted = ''.join(sorted(_str))
            if _sorted not in groups:
                groups[_sorted] = [_str]
            else:
                groups[_sorted].append(_str)
        return list(groups.values())

    def solve2(self, strs):
        pass

    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        return self.solve1(strs)

File: cpp/49.group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
	  unordered_map<string, vector<int> > st2pos;
	  vector<vector<string> > res;
	  int n = strs.size();
	  for(int i = 0 ; i < n; i++){
		  auto cpy = strs[i];
		  sort(cpy.begin(), cpy.end());
		  st2pos[cpy].push_back(i);
	  } 
	  for(auto [_, pos]:st2pos){
		  vector<string> tmp;
		  for(auto idx:pos){
			  tmp.push_back(strs[idx]);
		  }
		  res.push_back(tmp);
	  }
	  return res;
    }
};
